common/src/main/java/com/gregtechceu/gtceu/api/item/tool/TreeFellingHelper.java

```diff
@@ -0,0 +1,80 @@
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.state.BlockState;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class TreeFellingHelper {
+
+    public void fellTree(ItemStack stack, Level level, BlockState origin, BlockPos originPos, LivingEntity miner) {
+        BlockPos.MutableBlockPos mutablePos = new BlockPos.MutableBlockPos();
+
+        Queue<BlockPos> checking = new ArrayDeque<>();
+        Set<BlockPos> visited = new ObjectOpenHashSet<>();
+
+        checking.add(originPos);
+
+        while (!checking.isEmpty()) {
+            BlockPos check = checking.remove();
+            if (check != originPos) {
+                visited.add(check);
+            }
+            for (int y = 0; y <= 1; y++) {
+                for (int x = -1; x <= 1; x++) {
+                    for (int z = -1; z <= 1; z++) {
+                        if (x != 0 || y != 0 || z != 0) {
+                            mutablePos.set(check.getX() + x, check.getY() + y, check.getZ() + z);
+                            if (!visited.contains(mutablePos)) {
+                                // Check that the found block matches the original block state, which is wood.
+                                if (origin.getBlock() == level.getBlockState(mutablePos).getBlock()) {
+                                    if (!checking.contains(mutablePos)) {
+                                        BlockPos immutablePos = mutablePos.immutable();
+                                        checking.add(immutablePos);
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        if (!visited.isEmpty() && level.getServer() != null) {
+            List<BlockPos> orderedBlocks = visited.stream()
+                    .sorted(Comparator.comparingInt(pos -> pos.getY() - originPos.getY()))
+                    .collect(Collectors.toCollection(LinkedList::new));
+
+            int durabilityLeft = stack.getMaxDamage() - stack.getDamageValue();
+
+            if (durabilityLeft < orderedBlocks.size()) {
+                orderedBlocks.subList(durabilityLeft, orderedBlocks.size()).clear();
+            }
+
+
+            stack.hurtAndBreak(orderedBlocks.size(), miner, p -> p.broadcastBreakEvent(EquipmentSlot.MAINHAND));
+
+            breakBlocksPerTick((ServerPlayer) miner, orderedBlocks, stack);
+            orderedBlocks = new LinkedList<>();
+        }
+    }
+
+    public void breakBlocksPerTick(ServerPlayer player, List<BlockPos> posList, ItemStack stack) {
+        for (int i = 0; i < posList.size(); i++) {
+
+            int delayTick = i * 2; // 1 block per 2 tick
+            BlockPos pos = posList.get(i);
+            TaskHandler.enqueueServerTask(player.getLevel(), () -> {
+                player.getLevel().destroyBlock(pos, true);
+            }, delayTick);
```

Comment:
after delayTick, you should always check the block is still the block you want to break before performing real actions.
Many people ignore this problem.